### DISCUSSION TITLE
redis-server available to redis-cli but not to spec test?

### DISCUSSION THREAD

**Discussion by 573** [community/community#25063] [Actions]
<p>I have set up a workflow that extracts a nix via cachix/install-nix-action@v6 and then in the run step spawns a nix-installed redis-server, sends a ping to that server and starts a nix-build using my project’s default.nix.</p>
<p>When approaching the spec tests my build fails though saying the redis-server might not be available (<a href="https://github.com/573/shorturls/commit/1a17a260a003dcf1dd745efdccb52f7e251d0cd9/checks?check_suite_id=364429218#step:6:25121" rel="nofollow noopener">https://github.com/573/shorturls/commit/1a17a260a003dcf1dd745efdccb52f7e251d0cd9/checks?check_suite_id=364429218#step:6:25121</a>)</p>

**Comment by Yanjingzhu** [comment]
<p>Can you run your tests successfully in your local machine? If it works, could you please set up a self-hosted agent to run your workflow? Will the self-hosted agent managed to work?  </p>
<p>Also, you could create an issue in <a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fcachix%2Fcachix-action%2Fissues&amp;data=02%7C01%7Cv-yjingz%40microsoft.com%7C7abb625b625f458c9e0f08d787735720%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637126801760821637&amp;sdata=s%2FMY65OZhcQXtELFmwGk%2BJ3GDFuA0wGuKCr07NqEv%2BM%3D&amp;reserved=0" rel="nofollow noopener">https://github.com/cachix/cachix-action/issues</a>.</p>

**Comment by 573** [comment]
<p>Yes, locally the tests are run well.</p>
<p>I even attempted - locally - to start the redis-server with a redis.conf file of mine to start it in unix sockets mode, which also works - locally and with minor changes to my client code.</p>
<p>When trying to replicate the unix sockets starting approach in github workflow by adding the “CMD override” to the services section / options (<a href="https://github.community/t/job-service-command/16477/4">see this discussion</a>) I got the same error as before. Maybe the CMD override simply doesn’t work as I thought.</p>
<p>What do you mean by self-hosted agent ?</p>
<p>I already got some feedback on doing integration-testing using nix over at <a href="https://discourse.nixos.org/t/how-can-i-spawn-a-redis-instance-within-a-nix-build/5155" rel="nofollow noopener">discourse</a> which lead me to believe that the only option I have for now with my sparse knowledge is to try and start redis-server that way I described above.</p>
<p>What do you think it is possible ?</p>
<p>Anyway, thank you for your feedback, didn’t see it soon enough, as I had notification not yet enabled.</p>

**Comment by Yanjingzhu** [comment]
<p>Sorry for the delay response.</p>
<p>Self-hosted agent is installed on your local machine. You could see more detail information in this document: <a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fhelp.github.com%2Fen%2Factions%2Fautomating-your-workflow-with-github-actions%2Fabout-self-hosted-runners&amp;data=02%7C01%7Cv-yjingz%40microsoft.com%7C32da05d8025e4e2c14e308d794e52187%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637141584138649537&amp;sdata=H2euu4X9fsEVaAc0gIzXNbicdqSvX3fRpvv74Ag0mEQ%3D&amp;reserved=0" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/about-self-hosted-runners</a></p>
<p>I was working on reproduce your issue in my self-hosted runner in the last few days. I install redis-server in my self-hosted machine and keep it running. But I still got this Newwork.Socket.connect error. I don’t think starting redis-server could resolve this issue.  </p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/2/2bdef6d550901dc75066640656ab5b1645e6978b.png" data-download-href="/uploads/short-url/6g6efnMX8zKvCbIRFjyY5fIPhA7.png?dl=1" title="socket error .png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181088629-3cf9fcab-cecb-4aa4-a897-1c4805905690.png" alt="socket error .png" data-base62-sha1="6g6efnMX8zKvCbIRFjyY5fIPhA7" width="690" height="295" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/2/2bdef6d550901dc75066640656ab5b1645e6978b_2_690x295.png, https://aws1.discourse-cdn.com/github/optimized/2X/2/2bdef6d550901dc75066640656ab5b1645e6978b_2_1035x442.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/2/2bdef6d550901dc75066640656ab5b1645e6978b_2_1380x590.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/2/2bdef6d550901dc75066640656ab5b1645e6978b_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">socket error .png</span><span class="informations">1566×671 43.2 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>I just saw that the last workflow run in your repo <a href="https://github.com/573/nix-github-integration-tests/actions" rel="nofollow noopener">573/nix-github-integration-tests</a> succeeded . Do you still meet the problem which you report in your original post?  </p>

**Comment by 573** [comment]
<p><a href="https://github.com/573/nix-github-integration-tests/actions" rel="nofollow noopener">573/nix-github-integration-tests</a> is another repo, I was trying to make use of <a href="https://github.com/NixOS/nixpkgs/tree/master/nixos/tests" rel="nofollow noopener">nixos/tests</a> because I initially thought this could lead to a solution which it does not as <a href="https://github.com/NixOS/nixpkgs/tree/master/nixos/tests" rel="nofollow noopener">nixos/tests</a> seems <a href="https://discourse.nixos.org/t/running-tests-faster-raw-not-within-qemu/5396" rel="nofollow noopener">reportedly</a> fitted to run service to service tests only.</p>
<p>I am investigating further in the direction given <a href="https://discourse.nixos.org/t/running-tests-faster-raw-not-within-qemu/5396" rel="nofollow noopener">there</a> and will inform you in case of a result soon.</p>
<p>As far as I can guess the error does seem to have nothing to do with github actions but more with nix not allowing this kind of integration testing approach.</p>

**Comment by 573** [comment, marked as answer]
<p>Hello <a href="/t5/user/viewprofilepage/user-id/52380">Yanjingzhu</a>,</p>
<p>for a solution to my problem described above pls see <a href="https://discourse.nixos.org/t/how-can-i-spawn-a-redis-instance-within-a-nix-build/5155/10?u=573" rel="nofollow noopener">https://discourse.nixos.org/t/how-can-i-spawn-a-redis-instance-within-a-nix-build/5155/10?u=573</a> or even <a href="https://github.com/573/shorturls/tree/a34b28e98639607bd0590cd3ec76ef719ba2cc05" rel="nofollow noopener">my repository</a>.</p>
<p>There is indeed nothing wrong with Github Actions, sorry for the fuzz.</p>
<p>What needed to be done was running the service to be tested against <a href="https://github.com/573/shorturls/blob/a34b28e98639607bd0590cd3ec76ef719ba2cc05/default.nix#L22" rel="nofollow noopener">within (ll. 22)</a> the isolated build.</p>

**Comment by Yanjingzhu** [comment]
<p>I am so glad to hear that you have resolved your issue. If you have any other Github Actions questions, welcome to github community forum again. </p>
